---
title: "Vincent + Raymond Second Attempt"
output: html_notebook
---

# Load required packages

SIAMCAT is used for analysis

```{r}
library(curatedMetagenomicData)
library(dplyr)
library(mia)
library(tidymodels)
library(stringr)
library(ggplot2)
library(dotwhisker)
library(biobroom)
```

# Filter cMD metadata table for Vincent and Raymond
```{r}
meta <- curatedMetagenomicData::sampleMetadata
meta$study_name %>% table()

submeta <- meta %>% filter(study_name=="VincentC_2016" | study_name=="RaymondF_2016")
submeta$study_name %>% table()
```

# Download data from cMD for matching metadata
Table should match the metadata table
```{r}
rayvince <- returnSamples(submeta, "relative_abundance")
colData(rayvince)$study_name %>% table()
```

# Proportions in cMD are presented as percentages. Need to be decimals. Divide by 100.
```{r}
assay(rayvince) <- assay(rayvince)/100
```

#Subset to just Raymond post-exposure cases

There should be 24 data points (18 cases, 6 controls)
```{r}
ray <- subset(rayvince, select =colData(rayvince)$study_name=="RaymondF_2016")
#ray <- subset(rayvince, select =colData(rayvince)$study_name=="RaymondF_2016" &
              #  colData(rayvince)$days_from_first_collection == 7)

colData(ray)$study_name %>% table()
```
```{r}
raytidy <- tidy(ray, addPheno = TRUE)
raytidy$study_condition <- raytidy$study_condition %>% as.factor()
raywide <- pivot_wider(raytidy, names_from = gene, values_from = value)
ray7only <- raywide[raywide$days_from_first_collection==7,] # 7 days from collection only
```


# Try it with just a plain regression
```{r}
rayglm <- glm(study_condition ~ `class:Bacteroidia`, data = rayclasswide,
              family=binomial(link="logit"))
tidy(rayglm)
```

# Compare to tidymodels regression results, results should be identical

```{r}

lm_mod <- logistic_reg() %>% set_engine("glm")
lm_fit <- lm_mod %>% fit(study_condition ~ `class:Bacteroidia`, data=rayclasswide)
tidy(lm_fit)
```

# Try an RF of 7 days post data collection observations only
It's really bad. Probably need to do some resampling and maybe not agglomerate to class
```{r}
# Select only the microbiome vars
ray7sub <- ray7only %>% select(c("study_condition", starts_with("k__")))
set.seed(732981)
ray7split <- initial_split(ray7sub, strata = study_condition)
ray7train <- training(ray7split)
ray7test <- testing(ray7split)

rf_mod <- 
  rand_forest(trees = 1000) %>% 
  set_engine("ranger") %>% 
  set_mode("classification")

rf_fit <- rf_mod %>% 
  fit(study_condition ~ ., data=ray7train)

rf_testing_pred <- 
  predict(rf_fit, ray7test) %>% 
  bind_cols(predict(rf_fit, ray7test, type = "prob")) %>% 
  bind_cols(ray7test %>% select(study_condition))
rf_testing_pred %>%                   # test set predictions
    roc_auc(truth = study_condition, estimate=.pred_cephalosporins)
rf_testing_pred %>%                   # test set predictions
  accuracy(truth = study_condition, .pred_class)
```

```{r}
folds <- vfold_cv(ray7train, v= 4, strata = study_condition)
rf_wf <- 
  workflow() %>%
  add_model(rf_mod) %>%
  add_formula(study_condition ~ .)
rf_fit_rs <- 
  rf_wf %>% 
  fit_resamples(folds)
collect_metrics(rf_fit_rs)
```



```{r}
new <- expand.grid(phylum.Proteobacteria = c(0,0.25,0.50,0.75),
                   class.Bacteroidia = c(0,0.25,0.50,0.75),
                   class.Coriobacteriia = c(0,0.25,0.50,0.75)
                   )
predict(lm_fit, new_data = new, type="prob")
```

