---
title: "Zaura Tidy Models"
output: html_notebook
---

# Load required packages

SIAMCAT is used for analysis

```{r}
library(curatedMetagenomicData)
library(dplyr)
library(mia)
library(tidymodels)
library(stringr)
library(ggplot2)
library(dotwhisker)
library(biobroom)
library(phyloseq)
```

```{r}
phyloZaura <- import_biom("../data/Zaura/taxa.biom")
zaura <- phyloZaura
#treeZaura <- read_tree("../data/Zaura/phylo/rooted_tree.nwk")
metaZaura <- read.csv("../data/Zaura/zaura metadata.csv")
row.names(metaZaura) <- metaZaura$NCBI_accession
#phy_tree(zaura) <- treeZaura
sample_data(zaura) <- metaZaura
```


# Get NCBI taxon ids
```{r}
library(taxizedb)
db_download_ncbi()
ncbi <- src_ncbi()
ncbi
```
```{r}
zauragenus <- subset_taxa(zaura, Rank6!="none")
zaurag <- tax_glom(zauragenus, "Rank6")
zauraprop <- transform_sample_counts(zaurag, function(x) x / sum(x))

#Get column from tax table
taxCol <- tax_table(zauraprop)[,6]
#Send column cell to name2taxid
firstid <- function(x){
  y <- name2taxid(x, out_type = "summary")[1,2]
  return(y)
}
taxList <- lapply(taxCol, firstid)
taxNCBI <- map_dfr(taxList, bind_rows)

convert2ncbi <- function(phylo){
  taxtable <- tax_table(phylo)
  ranks <- dim(taxtable)[2]
  taxa <- dim(taxtable)[1]
  ncbis <- data.frame(num=1:taxa)
  for (x in 1:6){
    taxCol <- taxtable[,x]
    taxList <- lapply(taxCol, firstid)
    taxNCBI <- map_dfr(taxList, bind_rows)
    ncbis <- cbind(ncbis, taxNCBI)
  }
  return(ncbis)
}

ncbiranks <- convert2ncbi(zauraprop)

rvgenera <- rownames(otu_table(phylorv)) %>% stringr::str_extract(pattern="\\d{1,}")
ncbiranks[,7] %in%  rvgenera %>% table()
newrownames <- data.frame(newname=ncbiranks[,7], tax_table(zauraprop))

newrownames <- newrownames %>% 
    mutate(rankname = coalesce(newname,Rank6))

#There's one duplicate so let's just change it for now
newrownames$rankname[newrownames$Rank6=="Christensenella"] <- "duplicate"

taxa_names(zauraprop) <- newrownames$rankname
taxa_names(phylorv) <- rvgenera
#Use [1,2] result and return it
#Make new tax table with results

```



# Convert OTU table to proportions and trim
```{r}
# Select only taxa on the genus level


zauragenus <- filter_taxa(zaura, taxon_ranks == "Rank6")
zauraFilt= filter_taxa(zaura, function(x) mean(x) <.01,TRUE)
rmtaxa = taxa_names(zauraFilt)
alltaxa = taxa_names(zaura)

myTaxa = alltaxa[!alltaxa %in% rmtaxa]

zauraPruned <- prune_taxa(myTaxa,zaura)




```

# Turn it into a TSE and tidy it
```{r}
Zauratse <- makeTreeSummarizedExperimentFromPhyloseq(zauraPruned)
Zauratidy <- tidy(Zauratse, addPheno = TRUE)
Zaurawide <- pivot_wider(Zauratidy, names_from = gene, values_from = value)
Zaurasub<- Zaurawide %>% select(c("TreatmentGroup", starts_with("seq")))
Zaurasub$TreatmentGroup <- factor(Zaurasub$TreatmentGroup)
```


```{r}
# Select only the microbiome vars

set.seed(732981)
Zaurasplit <- initial_split(Zaurasub)
Zauratrain <- training(Zaurasplit)
Zauratest <- testing(Zaurasplit)

rf_mod <- 
  rand_forest(trees = 500) %>% 
  set_engine("ranger") %>% 
  set_mode("classification")

rf_fit <- rf_mod %>% 
  fit(TreatmentGroup ~ ., data=Zauratrain)

rf_testing_pred <- 
  predict(rf_fit, Zauratest) %>% 
  bind_cols(predict(rf_fit, Zauratest, type = "prob")) %>% 
  bind_cols(Zauratest %>% select(TreatmentGroup))
#rf_testing_pred %>%                   # test set predictions
#    roc_auc(truth = study_condition, estimate=.pred_amoxicillin)
rf_testing_pred %>%                   # test set predictions
  accuracy(truth = TreatmentGroup, .pred_class)
```

```{r}
folds <- vfold_cv(Zauratrain, v= 6, strata=TreatmentGroup, repeats=2)
rf_wf <- 
  workflow() %>%
  add_model(rf_mod) %>%
  add_formula(TreatmentGroup ~ .)
rf_fit_rs <- 
  rf_wf %>% 
  fit_resamples(folds)
collect_metrics(rf_fit_rs)
```


