---
title: "Vatanen/DIABIMMUNE Analysis"
output: html_notebook
---


```{r}
library(readxl)
library(curatedMetagenomicData)
library(dplyr)
library(stringr)
```


# Download taxonomic profiles and metadata
```{r}
download.file("https://diabimmune.broadinstitute.org/diabimmune/uploads/attachments/92/all.relative_abundance.txt", "data/vatanentaxprof.txt", method="wget")
download.file("https://diabimmune.broadinstitute.org/diabimmune/uploads/attachments/94/aad0917.SuppTable1.xls", "data/vatanenmeta.xls", method="wget")

```

```{r}
metaphlanToPhyloseq <- function(
  tax,
  metadat=NULL,
  simplenames=TRUE,
  roundtointeger=FALSE,
  split="|"){
  ## tax is a matrix or data.frame with the table of taxonomic abundances, rows are taxa, columns are samples
  ## metadat is an optional data.frame of specimen metadata, rows are samples, columns are variables
  ## if simplenames=TRUE, use only the most detailed level of taxa names in the final object
  ## if roundtointeger=TRUE, values will be rounded to the nearest integer
  xnames = rownames(tax)
  shortnames = gsub(paste0(".+\\", split), "", xnames)
  if(simplenames){
    rownames(tax) = shortnames
  }
  if(roundtointeger){
    tax = round(tax * 1e4)
  }
  x2 = strsplit(xnames, split=split, fixed=TRUE)
  taxmat = matrix(NA, ncol=max(sapply(x2, length)), nrow=length(x2))
  colnames(taxmat) = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "Strain")[1:ncol(taxmat)]
  rownames(taxmat) = rownames(tax)
  for (i in 1:nrow(taxmat)){
    taxmat[i, 1:length(x2[[i]])] <- x2[[i]]
  }
  taxmat = gsub("[a-z]__", "", taxmat)
  taxmat = phyloseq::tax_table(taxmat)
  otutab = phyloseq::otu_table(tax, taxa_are_rows=TRUE)
  if(is.null(metadat)){
    res = phyloseq::phyloseq(taxmat, otutab)
  }else{
    res = phyloseq::phyloseq(taxmat, otutab, phyloseq::sample_data(metadat))
  }
  return(res)
}
```


```{r}
vatMeta <- read_xls("~/aim3/data/vatanenmeta.xls")
vatTax <- read.table("~/aim3/data/vatanentaxprof.txt", header = TRUE, row.names=1)
vatAbx <- read.csv("~/aim3/data/diabimmune.csv", header = TRUE)
colnames(vatAbx) <- c("sample", "abx_since_last", "class")

```

# Looks like we gotta fix the sample names in vatTax

```{r}
newcols <- colnames(vatTax) %>% str_extract(pattern="\\w\\w\\w\\w\\w\\w\\w[_]\\w+[.]\\w")
parts <- newcols %>% str_extract(pattern="\\w\\w\\w\\w\\w\\w\\w")
assess <- newcols %>% str_extract(pattern="\\d+[.]\\w")
long <- data.frame(subject=parts, Assessment=assess, sample=newcols)
```

```{r}
vatMetaLong <- as.data.frame(right_join(vatMeta, long))
vatMetaFinal <- inner_join(vatAbx, vatMetaLong)
rownames(vatMetaFinal) <- vatMetaFinal$sample
colnames(vatTax) <- newcols
```

```{r}
phyloVat <- metaphlanToPhyloseq(vatTax, metadat=vatMetaFinal)
```

```{r}
library(curatedMetagenomicData)
library(dplyr)
library(mia)
library(SIAMCAT)
library(stringr)
```

```{r}
label.vat <- create.label(meta=sample_data(phyloVat), label="abx_since_last", case="Yes")
sc.obj <- siamcat(feat=otu_table(phyloVat), label=label.vat, meta=sample_data(phyloVat))
```

 Filter by abundance

Cutoff is set at 0.001. Might want to experiment with this.

```{r}
sc.filt <- filter.features(sc.obj, filter.method = "abundance", cutoff = 0.001)
```

# Normalization

I'm guessing due to ranked unit normalization this is using Rankits

```{r}
sc.norm <- normalize.features(
  sc.filt,
  norm.method = "rank.unit"
  )
```

# Prepare to run machine learning task

Split the data, not sure the difference between folds and resamples and how to select the right number--right now I just picked numbers that didn't give an error message

```{r}

sc.split <-  create.data.split(
  sc.norm,
  num.folds = 6,
  num.resample = 2,
  inseparable = "sample"
)

```

# Random Forest

Lasso also seems to work--others gave weird results (weird ROCs)

```{r}

sc.mod <- train.model(
  sc.split,
  method = "randomForest"
)

sc.pred <- make.predictions(sc.mod)
pred_matrix <- pred_matrix(sc.pred)

sc.eval <- evaluate.predictions(sc.pred)
model.evaluation.plot(sc.eval)

```


